src/core/data/ndarray/ndarray_vector.h

```diff
@@ -5,12 +5,216 @@
+                }
+
+                if constexpr (dim > 2)
+                {
+                    if (nCells_[2] != source.nCells_[2])
+                    {
+                        throw std::runtime_error(
+                            "Error NdArrayVector cannot be assigned, incompatible sizes");
+                    }
+                }
+            }
+
+            this->data_ = source.data_;
+            return *this;
+        }
+
+        NdArrayVector& operator=(NdArrayVector&& source)
+        {
+            if (nCells_[0] != source.nCells_[0])
+            {
+                throw std::runtime_error(
+                    "Error NdArrayVector cannot be assigned, incompatible sizes");
+            }
+
+            if constexpr (dim > 1)
+            {
+                if (nCells_[1] != source.nCells_[1])
+                {
+                    throw std::runtime_error(
+                        "Error NdArrayVector cannot be assigned, incompatible sizes");
+                }
+
+                if constexpr (dim > 2)
+                {
+                    if (nCells_[2] != source.nCells_[2])
+                    {
+                        throw std::runtime_error(
+                            "Error NdArrayVector cannot be assigned, incompatible sizes");
+                    }
+                }
+            }
+
+            this->data_ = std::move(source.data_);
+            return *this;
+        }
+
+
+        template<typename... Indexes>
+        DataType& operator()(Indexes... indexes)
+        {
+            auto params = std::tuple<Indexes...>{indexes...};
+            static_assert(sizeof...(Indexes) == dim);
+
+            if constexpr (dim == 1)
+            {
+                auto i = std::get<0>(params);
+
+                static_assert(std::is_unsigned<decltype(i)>::value);
+
+                return this->data_[i];
+            }
+
+            if constexpr (dim == 2)
+            {
+                auto i = std::get<0>(params);
+                auto j = std::get<1>(params);
+
+                static_assert(std::is_unsigned<decltype(i)>::value);
+                static_assert(std::is_unsigned<decltype(j)>::value);
+
+                return this->data_[j + i * nCells_[1]];
+            }
+
+            if constexpr (dim == 3)
+            {
+                auto i = std::get<0>(params);
+                auto j = std::get<1>(params);
+                auto k = std::get<2>(params);
+
+                static_assert(std::is_unsigned<decltype(i)>::value);
+                static_assert(std::is_unsigned<decltype(j)>::value);
+                static_assert(std::is_unsigned<decltype(k)>::value);
+
+                return this->data_[k + j * nCells_[2] + i * nCells_[1] * nCells_[2]];
+            }
+
+            if constexpr (dim != 1 && dim != 2 && dim != 3)
+            {
+                return 0;
+            }
+        }
+
+        template<typename... Indexes>
+        DataType const& operator()(Indexes... indexes) const
```

Comment:
https://github.com/PHAREHUB/PHARE/blob/master/src/core/data/ions/ion_population/ion_population.h#L167

this can be used to have only one real function declaration

@nicolasaunai what do you think?

Reply:
probably worth doing in that case would spare many lines

Reply:
I will also need your help on this one,